File: R/R/versions.R
###############################################
## Versions
## Updates the incoming LiPD to the most recent
## LiPD standards
###############################################


#' Add createdBy key to metdata
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @return list d: Metadata
add_created_by <- function(d){
  if(!("createdBy" %in% names(d))){
    d[["createdBy"]] <- "unknown"
  }
  return(d)
}

#' Check what version of LiPD this file is using. If none is found, assume it's using version 1.0
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @return list tmp: Version number and meta
get_lipd_version <- function(d){
  version <- NULL
  keys <- c("lipdVersion", "liPDVersion", "LiPDVersion")
  for (i in 1:length(keys)){
    key <- keys[[i]]
    if (key %in% names(d)){
      version <- d[[key]]
      d[[key]] <- NULL
    }
  }
  version <- as.numeric(version)
  if (isNullOb(version) || is.na(version)){
    # make a prompt that asks if they know what lipd version this file is. 
    yn <- readline("I didn't find a LiPD version number for this file. Do you know the version number? (y/n)")
    if(yn == "y"){
      vn <- readline("Enter the version number (1.0, 1.1, 1.2, 1.3): ")
      if(vn %in% c(1, 1.0, 1.1, 1.2, 1.3)){
        version <- vn
      } else {
        stop("That's not a valid response. Please try writeLipd again.")
      }
    } else if(yn == "n"){
      print("I'll assume this is a current version (v1.3) file, but that may be an incorrect assumption. Please keep a backup!")
      version <- 1.3
    }
  }
  else if (!(version %in% c(1, 1.0, 1.1, 1.2, 1.3))){
    print(sprintf("LiPD version is invalid: %s", version))
  }
  tmp <- list()
  d[["lipdVersion"]] <- version
  tmp[["meta"]] <- d
  tmp[["version"]] <- version
  return(tmp)
}


#' Use the current version number to determine where to start updating from. Use "chain versioning" to make it
#' modular. If a file is a few versions behind, convert to EACH version until reaching current. If a file is one
#' version behind, it will only convert once to the newest.
#' @export
#' @keywords internal
#' @param d Metadata
#' @return d Metadata
update_lipd_version <- function(d){
  tryCatch({
    # Get the lipd version number, and updated meta
    tmp <- get_lipd_version(d)
    version <- tmp[["version"]]
    d <- tmp[["meta"]]
    
    # Update from (N/A or 1.0) to 1.1
    if (version == 1.0 || version == "1.0"){
      d = update_lipd_v1_1(d)
      version <- 1.1
      # print("updating to 1.1")
    }
    
    # Update from 1.1 to 1.2
    if (version == 1.1 || version == "1.1"){
      d = update_lipd_v1_2(d)
      version = 1.2
      # print("updating to 1.2")
    }
    
    # Update from 1.2 to 1.3
    if(version == 1.2 || version == "1.2"){
      d = update_lipd_v1_3(d)
      version = 1.3
      # print("updating to 1.3")
    }
  }, error=function(cond){
    print(paste0("Error: update_lipd_version: v", version, ": ", cond))
  })
    return(d)
}


#' Update LiPD v1.0 to v1.1
#' - chronData entry is a list that allows multiple tables
#' - paleoData entry is a list that allows multiple tables
#' - chronData now allows measurement, model, summary, ensemble, calibratedAges tables
#' - Added 'lipdVersion' key
#' @export
#' @keywords internal
#' @param d Metadata
#' @return d Metadata
update_lipd_v1_1 <- function(d){
  # Cannot v1.0 files, and I'm pretty sure there are less than 5 files in existence.
  return(d)
}

#' Update LiPD v1.1 to v1.2
#' - Added NOAA compatible keys : maxYear, minYear, originalDataURL, WDCPaleoURL, etc
#' - 'calibratedAges' key is now 'distribution' (handled in update_lipd_v1_3_keys instead)
#' - paleoData structure mirrors chronData. Allows measurement, model, summary, ensemble,
#'   distribution tables
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @return list d: Metadata
update_lipd_v1_2 <- function(d){
  if("paleoData" %in% names(d)){
    d <- update_lipd_v1_2_section(d, "paleoData")
  }
  if("chronData" %in% names(d)){
    d <- update_lipd_v1_2_section(d, "chronData")
  }
  d[["lipdVersion"]] <- 1.2
  return(d)
}

#' Update LiPD v1.1 to v1.2 - one section
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @param char pc: paleoData or chronData
#' @return list d: Metadata
update_lipd_v1_2_section <- function(d, pc){
  if(!isNullOb(d[[pc]])){
    # Create the structure that we want. 
    tmp <- list()
    tmp[["measurementTable"]] <- list()
    # Append the existing tables as mesurement tables in the new structure
    for(i in 1:length(d[[pc]])){
      tmp[["measurementTable"]][[i]] <- d[[pc]][[i]]
    }
    # Overwrite the old structure with the new structure
    d[[pc]] <- list()
    d[[pc]][[1]] <- tmp
  }
  return(d)
}

#' Update LiPD v1.2 to v1.3
#' - Added 'createdBy' key
#' - Top-level folder inside LiPD archives are named "bag". (No longer <datasetname>)
#' - .jsonld file is now generically named 'metadata.jsonld' (No longer <datasetname>.lpd )
#' - All "paleo" and "chron" prefixes are removed from "paleoMeasurementTable", "paleoModel", etc.
#' - Merge isotopeInterpretation and climateInterpretation into "interpretation" block
#' - ensemble table entry is a list that allows multiple tables
#' - summary table entry is a list that allows multiple tables
#' @export
#' @keywords internal
#' @param d Metadata
#' @return d Metadata
update_lipd_v1_3 <- function(d){
  tryCatch({
    d <- update_lipd_v1_3_keys(d)
    d <- update_lipd_v1_3_structure(d)
    d <- add_created_by(d)
  }, error=function(cond){
    print(paste0("Error: update_lipd_v1_3: ", cond))
  })
  d[["lipdVersion"]] <- 1.3
  return(d)
}

#' Update v1.2 keys to v1.3 keys: recursive
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @return list d: Metadata
update_lipd_v1_3_keys <- function(d){
  v12keys <- c("paleoMeasurementTable", "chronMeasurementTable", "paleoModel", "chronModel", "paleoDataMD5", "chronDataMD5", "paleoEnsembleMD5",  
               "chronEnsembleMD5", "paleoEnsembleTableMD5", "chronEnsembleTableMD5", "paleoMeasurementTableMD5", "chronMeasurementTableMD5", "name",
               "calibratedAges")
  v13keys <- c("measurementTable", "measurementTable", "model", "model", "dataMD5", "dataMD5", "tableMD5", "tableMD5", "tableMD5", "tableMD5",
               "tableMD5", "tableMD5", "tableName", "distributionTable")
  v13keymap <- setNames(as.list(v13keys), v12keys)
  v12keys_rm <- c("chronTableName", "paleoTableName", "paleoDataTableName", "chronDataTableName",
                  "chronMeasurementTableName", "paleoMeasurementTableName")
  
  # Keys that are in this list
  keys <- names(d)
  
  # For any lists that are indexed by name 
  if(!isNullOb(keys) && !is.na(keys)){
    for(i in 1:length(keys)){
      old_key <- keys[[i]]
      # Dive down first
      if(typeof(d[[old_key]]) == "list"){
        d[[old_key]] <- update_lipd_v1_3_keys(d[[old_key]])
      } 
      # When you bubble back up, then check if this key should be switched
      if(old_key %in% v12keys){
        # Set the old
        new_key <- v13keymap[[old_key]]
        d[[new_key]] <- d[[old_key]]
        # Remove the old key
        d[[old_key]] <- NULL
      } else if(old_key %in% v12keys_rm){
        d[[old_key]] <- NULL
      }
    }
  } else {
    # For any lists that are indexed by number
    if(typeof(d)=="list"){
      for(i in 1:length(d)){
        tryCatch({
          d[[i]] <- update_lipd_v1_3_keys(d[[i]])
        }, error=function(cond){
          print(paste0("Error: update_lipd_v1_3_keys: ", cond))
        })
      }
    }
  }
  return(d)
}

#' Update the structure for summary and ensemble tables
#' @export
#' @keywords internal
#' @param d Metadata
#' @return d Metadata
update_lipd_v1_3_structure <- function(d){
  
  # Tables that need changing
  change <- c("ensembleTable", "summaryTable")
  interp <- c("isotopeInterpretation", "climateInterpretation")
  
  # Keys in this data
  keys <- names(d)
  
  # For any lists that are indexed by name 
  if(!isNullOb(keys) && !is.na(keys)){
    for(i in 1:length(keys)){
      curr_key <- keys[[i]]
      # Dive down first
      if(typeof(d[[curr_key]]) == "list"){
        d[[curr_key]] <- update_lipd_v1_3_structure(d[[curr_key]])
      } 
      # When you bubble back up, then check if this key should be switched
      if(curr_key %in% change){
        # Found a table. Do the 'ol switcheroo
        tmp <- d[[curr_key]]
        d[[curr_key]] <- list()
        d[[curr_key]][[1]] <- tmp
      } else if (curr_key %in% interp){
        d <- merge_interpretations(d, curr_key)
      }
    }
  } else {
    # For any lists that are indexed by number
    if(typeof(d)=="list"){
      for(i in 1:length(d)){
        tryCatch({
          d[[i]] <- update_lipd_v1_3_structure(d[[i]])
        }, error=function(cond){
          print(paste0("Error: update_lipd_v1_3_structure: ", cond))
        })
      }
    }
  }
  return(d)
}


#' Merge the old interpretation fields into the new, combined interpretation field
#' @export
#' @keywords internal
#' @param list d: Metadata
#' @param char key: climateInterpretation or isotopeInterpretation
#' @return list d: Metadata
merge_interpretations <- function(d, key){
  scope <- gsub("Interpretation", "", key)
  if (!("interpretation" %in% names(d))){
    d[["interpretation"]] <- list()
  }
  pos <- length(d[["interpretation"]]) + 1
  d[["interpretation"]][[pos]] <- d[[key]]
  d[["interpretation"]][[pos]][["scope"]] <-scope
  d[[key]] <- NULL
  return(d)
}

# DEPRECATED CODE

#' #' Convert LiPD version structure whenever necessary
#' #' @export
#' #' @keywords internal
#' #' @param d Metadata
#' #' @return d Metadata
#' convertVersion <- function(d){
#'   # Check which version this LiPD file is
#'   d <- check_version(d)
#'   # check and convert any data frames into lists
#'   d <- convert_dataframes_list(d)
#'   return(d)
#' }
#' 
#' #' Check / convert and fixed data frames into scalable lists
#' #' @export
#' #' @keywords internal
#' #' @param d LiPD metadata
#' #' @return d Modified LiPD metadata
#' convert_dataframes_list <- function(d){
#'   # convert single entries to lists. matching structure to 1.2
#'   d <- convert_single_multiple(d, "paleoData")
#'   d <- convert_single_multiple(d, "chronData")
#' 
#'   return(d)
#' }
#' 
#' #' Convert from a single fixed table, to a multiple scalable table
#' #' (LiPD Verison 1.1 to 1.2 change)
#' #' @export
#' #' @keywords internal
#' #' @param d LiPD metadata
#' #' @return d Modified LiPD metadata
#' convert_single_multiple <- function(d, pc){
#' 
#'   # data frame?
#'   dat <- hasData(d, pc)
#' 
#'   # proceed of section exists
#'   if (!is.null(dat)){
#' 
#'     if (is.data.frame(dat)){
#'       d[[pc]] <- list()
#'       d[[pc]][[1]] <- as.list(dat)
#'     }
#'     # multiples?
#'     dat <- hasData(d[[pc]], 1)
#'     # convert to multiples
#'     if (is.null(dat)){
#'       d[[pc]] <- list()
#'       d[[pc]][[1]] <- dat
#'     } # END PALEODATA
#' 
#'     # loop
#'     for (i in 1:length(d[[pc]])){
#' 
#'       # PALEODATA[[i]]
#'       # data frame?
#'       dat <- hasData(d[[pc]], i)
#'       if (is.data.frame(!is.null(dat))){
#'           d[[pc]][[i]] <- as.list(dat)
#'       }
#'       # MEAS + MODEL
#'       # table exists ?
#'       # d$paleoData[[i]]$paleoMeasurementTable
#'       dat.meas <- hasData(d[[pc]][[i]], "measurementTable")
#' 
#'       # table exists ?
#'       # d$paleoData[[i]]$paleoModel
#'       dat.model <- hasData(d[[pc]][[i]], "model")
#' 
#'       # tables do not exist.
#'       # make a meas table
#'       if (is.null(dat.meas) & is.null(dat.model)){
#'         tmp <- d[[pc]][[i]]
#'         d[[pc]][[i]] <- list()
#'         d[[pc]][[i]][["measurementTable"]] <- list()
#'         d[[pc]][[i]][["measurementTable"]][[1]] <- tmp
#'       }  # end meas and model
#' 
#'       # DIRECT
#'       # multiples ?
#'       # d$paleoData[[i]]$paleoMeasurementTable$columns
#'       dat <- hasData(d[[pc]][[i]][["measurementTable"]], "columns")
#'       # convert to multiples
#'       # d$paleoData[[i]]$paleoMeasurementTable
#'       if (!is.null(dat)){
#'         tmp <- d[[pc]][[i]][["measurementTable"]]
#'         d[[pc]][[i]][["measurementTable"]] <- list()
#'         d[[pc]][[i]][["measurementTable"]][[1]] <- tmp
#'       } # end direct data
#' 
#'       # MEASUREMENT
#'       # paleoData[[i]]paleoMeasurementTable
#'       # data frame ?
#'       dat <- hasData(d[[pc]][[i]], "measurementTable")
#'       if (is.data.frame(!is.null(dat))){
#'         d[[pc]][[i]][["measurementTable"]] <- as.list(dat)
#'       }
#'       # multiples ?
#'       if (!is.null(dat)){
#'         dat <- hasData(d[[pc]][[1]][["measurementTable"]], 1)
#'         # convert to multiples
#'         # d$paleoData[[i]]$paleoMeasurementTable[[j]]
#'         if (is.null(dat)){
#'           tmp <- d[[pc]][[1]][["measurementTable"]]
#'           d[[pc]][[1]][["measurementTable"]] <- list()
#'           d[[pc]][[1]][["measurementTable"]][[1]] <- tmp
#'         } # END MEASUREMENT
#' 
#'         # loop
#'         for (j in 1:length(d[[pc]][[i]][["measurementTable"]])){
#' 
#'           # MEASUREMENT[[j]]
#'           # paleoData[[i]]paleoMeasurementTable[[j]]
#'           # data frame?
#'           dat <- hasData(d[[pc]][[i]][["measurementTable"]], j)
#'           if (is.data.frame(!is.null(dat))){
#'             d[[pc]][[i]][["measurementTable"]][[j]] <- as.list(dat)
#'           } # END MEASUREMENT[[j]]
#' 
#'         } # end meas loop
#'       } # end meas exists
#' 
#'       # continue if Model table present
#'       if (length(d[[pc]][[i]][["model"]]) > 0){
#'         # MODEL
#'         # paleoData[[i]]paleoModel
#'         # data frame ?
#'         dat <- hasData( d[[pc]][[i]], "model")
#'         if (is.data.frame(!is.null(dat))){
#'           d[[pc]][[i]][["model"]] <- as.list(dat)
#'         }
#'         # multiples ?
#'         # convert to multiples
#'         if (!is.null(dat)){
#'           dat <- hasData(d[[pc]][[1]][["model"]], 1)
#'           if (is.null(dat)){
#'             tmp <- d[[pc]][[1]][["model"]]
#'             d[[pc]][[1]][["model"]] <- list()
#'             d[[pc]][[1]][["model"]][[1]] <- tmp
#'           } # END MODEL
#' 
#'           # loop
#'           for (j in 1:length(d[[pc]][[i]][["model"]])){
#' 
#'             # MODEL[[j]]
#'             # paleoModel[[j]]
#'             # data frame ?
#'             dat <- hasData(d[[pc]][[i]][["model"]], j)
#'             if (is.data.frame(!is.null(dat))){
#'               d[[pc]][[i]][["model"]][[j]] <- as.list(dat)
#'             }
#' 
#'             # SUMMARY
#'             # paleoModel[[j]]$summaryTable
#'             # data frame ?
#'             dat <- hasData(d[[pc]][[i]][["model"]][[j]], "summaryTable")
#'             if (is.data.frame(!is.null(dat))){
#'               d[[pc]][[i]][["model"]][[j]][["summaryTable"]] <- as.list(dat)
#'             }
#' 
#'             # ENSEMBLE
#'             # paleoModel[[j]]$ensembleTable
#'             # data frame ?
#'             dat <- hasData(d[[pc]][[i]][["model"]][[j]], "ensembleTable")
#'             if (is.data.frame(!is.null(dat))){
#'               d[[pc]][[i]][["model"]][[j]][["ensembleTable"]] <- as.list(dat)
#'             }
#' 
#'             # DISTRIBUTION
#'             # paleoModel[[j]]$distributionTable
#'             dat <- hasData(d[[pc]][[i]][["model"]][[j]], "distributionTable")
#'             if (is.data.frame(!is.null(dat))){
#'               d[[pc]][[i]][["model"]][[j]][["distributionTable"]] <- as.list(dat)
#'             } # end dist
#' 
#'             # multiples ?
#'             # convert to multiples
#'             if (!is.null(dat)){
#'               dat <- hasData(d[[pc]][[1]][["model"]][[j]][["distributionTable"]], 1)
#'               if (is.null(dat) & !is.null(dat)){
#'                 tmp <- d[[pc]][[i]][["model"]][[j]][["distributionTable"]]
#'                 d[[pc]][[i]][["model"]][[j]][["distributionTable"]] <- list()
#'                 d[[pc]][[i]][["model"]][[j]][["distributionTable"]][[1]] <- tmp
#'               } # end dist 1
#' 
#'               # loop
#'               for (k in 1:length(d[[pc]][[i]][["model"]][[j]][["distributionTable"]])){
#' 
#'                 # DISTRIBUTION[[k]]
#'                 dat <- hasData(d[[pc]][[i]][["model"]][[j]][["distributionTable"]], k)
#'                 if (is.data.frame(!is.null(dat))){
#'                   d[[pc]][[i]][["model"]][[j]][["distributionTable"]][[k]] <- as.list(dat)
#'                 } # END DISTRIBUTION[[k]]
#' 
#'               } # end dist loop
#' 
#'             } # end dist exists
#' 
#'           } # end models
#' 
#'         } # end model exists
#' 
#'       } # end if
#' 
#'     } # end section
#' 
#'   } # end if section
#' 
#'   # change the LiPDVersion value to 1.2
#'   d[["lipdVersion"]] <- 1.2
#'   return(d)
#' }

File: Python/lipd/versions.py
from .alternates import VER_1_3
from .loggers import create_logger
from .misc import is_one_dataset

logger_versions = create_logger("versions")


def fix_pubYear(D):
    """
    Move the DOI from the BibJson location, to our desired pub top-level location under "doi"

    Remove 'pubYear' keys and reassign to 'year' if necessary.

    :param   dict D: Metadata
    :return  dict D: Metadata, with data modified where applicable
    """
    if is_one_dataset:
        if "pub" in D:
            # Loop for each publication entry
            for pub in D["pub"]:
                if "pubYear" in pub:
                    if "year" not in pub:
                        pub["year"] = pub["pubYear"]
                    del pub["pubYear"]

    else:
        # Loop for each dataset if D is multiple datasets
        for idx, L in enumerate(D):
            D[idx] = fix_pub(L)

    return D


def fix_doi(L):
    """
    'pubYear' keys need to be switched to 'year' where applicable

    DOIs are commonly stored in the BibJson format under "identifier". We want to move these to the root
    of the publication under "doi". Make the reassignments necessary and also remove duplicate and unwanted
    Doi/DOI keys.

    :param  dict L:  Metadata
    :return dict L:  Metdata
    """
    # Keys that we don't want. Reassign data to 'doi' key
    _dois = ["DOI", "Doi"]
    # Loop for each publication entry
    try:
        if "pub" in L:
            for pub in L["pub"]:
                try:
                    # Is there an identifier in this publication?
                    if "identifier" in pub:
                        # Attempt to grab the doi from the id location. If it doesn't work, we'll catch the error.
                        _identifier = pub["identifier"][0]["id"]
                        # Got identifier. Is there a valid string here?
                        if _identifier:
                            # Reassign the doi to the publication root 'doi' key
                            pub["doi"] = _identifier
                            # Delete the identifier key and data
                            del pub["identifier"]
                except Exception:
                    # Catch the KeyError, and continue on normally.
                    pass

                # Check for each doi key that we don't want
                for _key in _dois:
                    # Is it in the publication entry?
                    if _key in pub:
                        #  Is there valid string data?
                        if pub[_key]:
                            # Reassign the doi to the 'doi' key
                            pub["doi"] = pub[_key]
                        # Delete the bad doi key
                        del pub[_key]
        else:
            L["pub"] = []
    except Exception as e:
        logger_versions.info("fix_doi: Publication error: {}".format(e))
    return L


def get_lipd_version(L):
    """
    Check what version of LiPD this file is using. If none is found, assume it's using version 1.0
    :param dict L: Metadata
    :return float:
    """
    version = 1.0
    _keys = ["LipdVersion", "LiPDVersion", "lipdVersion", "liPDVersion"]
    for _key in _keys:
        if _key in L:
            version = L[_key]
            # Cast the version number to a float
            try:
                version = float(version)
            except AttributeError:
                # If the casting failed, then something is wrong with the key so assume version is 1.0
                version = 1.0
            L.pop(_key)
    return L, version


def _merge_interpretations(d):
    """

    :param d:
    :return:
    """
    _tmp = []
    try:
        # Now loop and aggregate the interpretation data into one list
        for k, v in d.items():
            if k in ["climateInterpretation", "isotopeInterpretation", "interpretation"]:
                # now add in the new data
                if isinstance(v, list):
                    for i in v:
                        _tmp.append(i)
                elif isinstance(v, dict):
                    _tmp.append(d[k])
        # Set the aggregate data into the interpretation key
        d["interpretation"] = _tmp

    except Exception as e:
        print("Error: merge_interpretations: {}".format(e))

    # Now remove the old interpretation keys
    for key in ["climateInterpretation", "isotopeInterpretation"]:
        try:
            d[key] = ""
        except Exception:
            pass

    return d


def update_lipd_version(L):
    """
    Metadata is indexed by number at this step.

    Use the current version number to determine where to start updating from. Use "chain versioning" to make it
    modular. If a file is a few versions behind, convert to EACH version until reaching current. If a file is one
    version behind, it will only convert once to the newest.
    :param dict L: Metadata
    :return dict d: Metadata
    """
    # Get the lipd version number.
    L, version = get_lipd_version(L)

    # Update from (N/A or 1.0) to 1.1
    if version in [1.0, "1.0"]:
        L = update_lipd_v1_1(L)
        version = 1.1

    # Update from 1.1 to 1.2
    if version in [1.1, "1.1"]:
        L = update_lipd_v1_2(L)
        version = 1.2
    if version in [1.2, "1.2"]:
        L = update_lipd_v1_3(L)
        version = 1.3

    L = fix_doi(L)
    L = fix_pubYear(L)
    L["lipdVersion"] = 1.3
    return L


def update_lipd_v1_1(d):
    """
    Update LiPD v1.0 to v1.1
    - chronData entry is a list that allows multiple tables
    - paleoData entry is a list that allows multiple tables
    - chronData now allows measurement, model, summary, modelTable, ensemble, calibratedAges tables
    - Added 'lipdVersion' key

    :param dict d: Metadata v1.0
    :return dict d: Metadata v1.1
    """
    logger_versions.info("enter update_lipd_v1_1")
    tmp_all = []

    try:
        # ChronData is the only structure update
        if "chronData" in d:
            # As of v1.1, ChronData should have an extra level of abstraction.
            # No longer shares the same structure of paleoData

            # If no measurement table, then make a measurement table list with the table as the entry
            for table in d["chronData"]:
                if "chronMeasurementTable" not in table:
                    tmp_all.append({"chronMeasurementTable": [table]})

                # If the table exists, but it is a dictionary, then turn it into a list with one entry
                elif "chronMeasurementTable" in table:
                    if isinstance(table["chronMeasurementTable"], dict):
                        tmp_all.append({"chronMeasurementTable": [table["chronMeasurementTable"]]})
            if tmp_all:
                d["chronData"] = tmp_all

        # Log that this is now a v1.1 structured file
        d["lipdVersion"] = 1.1
    except Exception as e:
        logger_versions.error("update_lipd_v1_1: Exception: {}".format(e))
    logger_versions.info("exit update_lipd_v1_1")
    return d


def update_lipd_v1_2(d):
    """
    Update LiPD v1.1 to v1.2
    - Added NOAA compatible keys : maxYear, minYear, originalDataURL, WDCPaleoURL, etc
    - 'calibratedAges' key is now 'distribution'
    - paleoData structure mirrors chronData. Allows measurement, model, summary, modelTable, ensemble,
        distribution tables
    :param dict d: Metadata v1.1
    :return dict d: Metadata v1.2
    """
    logger_versions.info("enter update_lipd_v1_2")
    tmp_all = []

    try:
        # PaleoData is the only structure update
        if "paleoData" in d:
            # As of 1.2, PaleoData should match the structure of v1.1 chronData.
            # There is an extra level of abstraction and room for models, ensembles, calibrations, etc.
            for table in d["paleoData"]:
                if "paleoMeasurementTable" not in table:
                    tmp_all.append({"paleoMeasurementTable": [table]})

                # If the table exists, but it is a dictionary, then turn it into a list with one entry
                elif "paleoMeasurementTable" in table:
                    if isinstance(table["paleoMeasurementTable"], dict):
                        tmp_all.append({"paleoMeasurementTable": [table["paleoMeasurementTable"]]})
            if tmp_all:
                d["paleoData"] = tmp_all
        # Log that this is now a v1.1 structured file
        d["lipdVersion"] = 1.2
    except Exception as e:
        logger_versions.error("update_lipd_v1_2: Exception: {}".format(e))

    logger_versions.info("exit update_lipd_v1_2")
    return d


def update_lipd_v1_3(d):
    """
    Update LiPD v1.2 to v1.3
    - Added 'createdBy' key
    - Top-level folder inside LiPD archives are named "bag". (No longer <datasetname>)
    - .jsonld file is now generically named 'metadata.jsonld' (No longer <datasetname>.lpd )
    - All "paleo" and "chron" prefixes are removed from "paleoMeasurementTable", "paleoModel", etc.
    - Merge isotopeInterpretation and climateInterpretation into "interpretation" block
    - ensemble table entry is a list that allows multiple tables
    - summary table entry is a list that allows multiple tables
    :param dict d: Metadata v1.2
    :return dict d: Metadata v1.3
    """
    # sub routine (recursive): changes all the key names and merges interpretation
    d = update_lipd_v1_3_names(d)
    # sub routine: changes ensemble and summary table structure
    d = update_lipd_v1_3_structure(d)
    d["lipdVersion"] = 1.3
    if "LiPDVersion" in d:
        del d["LiPDVersion"]
    return d


def update_lipd_v1_3_names(d):
    """
    Update the key names and merge interpretation data
    :param dict d: Metadata
    :return dict d: Metadata
    """
    try:
        if isinstance(d, dict):
            for k, v in d.items():
                # dive down for dictionaries
                d[k] = update_lipd_v1_3_names(v)

                # see if the key is in the remove list
                if k in VER_1_3["swap"]:
                    # replace the key in the dictionary
                    _key_swap = VER_1_3["swap"][k]
                    d[_key_swap] = d.pop(k)
                elif k in VER_1_3["tables"]:
                    d[k] = ""
            for _key in ["climateInterpretation", "isotopeInterpretation"]:
                if _key in d:
                    d = _merge_interpretations(d)

        elif isinstance(d, list):
            # dive down for lists
            for idx, i in enumerate(d):
                d[idx] = update_lipd_v1_3_names(i)
    except Exception as e:
        print("Error: Unable to update file to LiPD v1.3: {}".format(e))
        logger_versions.error("update_lipd_v1_3_names: Exception: {}".format(e))
    return d


def update_lipd_v1_3_structure(d):
    """
    Update the structure for summary and ensemble tables
    :param dict d: Metadata
    :return dict d: Metadata
    """
    for key in ["paleoData", "chronData"]:
        if key in d:
            for entry1 in d[key]:
                if "model" in entry1:
                    for entry2 in entry1["model"]:
                        for key_table in ["summaryTable", "ensembleTable"]:
                            if key_table in entry2:
                                if isinstance(entry2[key_table], dict):
                                    try:
                                        _tmp = entry2[key_table]
                                        entry2[key_table] = []
                                        entry2[key_table].append(_tmp)
                                    except Exception as e:
                                        logger_versions.error("update_lipd_v1_3_structure: Exception: {}".format(e))
    return d


